infra/Pos/Subscription/Common.hs

```diff
@@ -86,24 +108,61 @@ subscriptionListener
+    recvLimited conv >>= \case
+        Just MsgSubscribe -> do
+            let peers = simplePeers [(nodeType, nodeId)]
+            bracket
+              (updatePeersBucket BucketSubscriptionListener (<> peers))
+              (\added -> when added $ do
+                void $ updatePeersBucket BucketSubscriptionListener (removePeer nodeId)
+                logDebug $ sformat ("subscriptionListener: removed "%shown) nodeId)
+              (\added -> when added $ do -- if not added, close the conversation
+                  logDebug $ sformat ("subscriptionListener: added "%shown) nodeId
+                  fix $ \loop -> recvLimited conv >>= \case
+                      Just MsgSubscribeKeepAlive -> do
+                          logDebug $ sformat
+                              ("subscriptionListener: received keep-alive from "%shown)
+                              nodeId
+                          loop
+                      msg -> logNotice $ expectedMsgFromGot MsgSubscribeKeepAlive
+                                                            nodeId msg)
+        msg -> logNotice $ expectedMsgFromGot MsgSubscribe nodeId msg
+  where
+    expectedMsgFromGot = sformat
+            ("subscriptionListener: expected "%shown%" from "%shown%
+             ", got "%shown%", closing the connection")
+
+-- | Version of subscriptionListener for MsgSubscribe1.
+subscriptionListener1
+    :: forall pack m.
+       (SubscriptionMode m)
+    => OQ.OutboundQ pack NodeId Bucket
+    -> NodeType
+    -> (ListenerSpec m, OutSpecs)
+subscriptionListener1 oq nodeType = listenerConv @Void oq $ \_ourVerInfo nodeId conv -> do
     mbMsg <- recvLimited conv
-    whenJust mbMsg $ \MsgSubscribe -> do
+    whenJust mbMsg $ \MsgSubscribe1 -> do
       let peers = simplePeers [(nodeType, nodeId)]
       bracket
-        (updatePeersBucket BucketSubscriptionListener (<> peers))
-        (\added -> when added $
-          void $ updatePeersBucket BucketSubscriptionListener (removePeer nodeId))
-        (\added -> when added $
-          void $ recvLimited conv) -- if not added, close the conversation
+          (updatePeersBucket BucketSubscriptionListener (<> peers))
+          (\added -> when added $ do
+              void $ updatePeersBucket BucketSubscriptionListener (removePeer nodeId)
+              logDebug $ sformat ("subscriptionListener1: removed "%shown) nodeId)
+          (\added -> when added $ do -- if not added, close the conversation
+              logDebug $ sformat ("subscriptionListener1: added "%shown) nodeId
+              void $ recvLimited conv)
 
 subscriptionListeners
     :: forall pack m.
        (SubscriptionMode m)
     => OQ.OutboundQ pack NodeId Bucket
     -> NodeType
     -> MkListeners m
-subscriptionListeners oq nodeType = constantListeners [subscriptionListener oq nodeType]
+subscriptionListeners oq nodeType = constantListeners
+    [ subscriptionListener  oq nodeType
+    , subscriptionListener1 oq nodeType
+    ]
```

Comment:
Or is this where the order is important?
Or both?

Reply:
Order doesn't matter here.

Reply:
Agree, no order required